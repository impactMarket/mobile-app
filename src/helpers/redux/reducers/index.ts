import { combineReducers } from 'redux';

import { appReducer } from './app';
import { authReducer } from './auth';
import { userReducer } from './user';
import { modalDonateReducer } from './view';

export default combineReducers({
    user: userReducer,
    auth: authReducer,
    app: appReducer,
    modalDonate: modalDonateReducer,
});
