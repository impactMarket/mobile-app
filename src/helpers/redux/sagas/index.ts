import { all, spawn } from 'typed-redux-saga';

import { startWatchingNetworkConnectivity } from './offline';
import stories from './stories';

export default function* rootSagas() {
    return yield all([stories]);
    // return yield all([spawn(startWatchingNetworkConnectivity), stories]);
}
