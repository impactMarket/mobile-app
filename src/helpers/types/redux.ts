import { ContractKit } from '@celo/contractkit';
import { Subscription } from '@unimodules/core';
import {
    SET_USER_CELO_INFO,
    SET_USER_INFO,
    SET_USER_EXCHANGE_RATE,
    // INIT_USER,
    SET_USER_WALLET_BALANCE,
    SET_USER_IS_BENEFICIARY,
    SET_USER_IS_BLOCKED,
    SET_USER_IS_SUSPECT,
    SET_USER_IS_COMMUNITY_MANAGER,
    SET_CELO_KIT,
    SET_APP_BENEFICIARY_HAS_ACCEPTED_TERMS,
    SET_APP_MANAGER_HAS_ACCEPTED_TERMS,
    SET_COMMUNITY_CONTRACT,
    SET_COMMUNITY,
    RESET_USER_APP,
    // RESET_NETWORK_APP,
    SET_PUSH_NOTIFICATION_TOKEN,
    SET_AUTH_TOKEN,
    SET_USER_LANGUAGE,
    SET_APP_SUSPECT_WRONG_DATETIME,
    SET_APP_FROM_WELCOME_SCREEN,
    SET_EXCHANGE_RATES,
    appAction,
    modalDonateAction,
    storiesAction,
} from 'helpers/constants';

import { ICommunitiesListStories, ICommunityStory } from './endpoints';
import { CommunityAttributes, UserAttributes } from './models';
import { IUserWallet } from './state';

// action
interface UserWalletAction {
    type: typeof SET_USER_CELO_INFO;
    payload: IUserWallet;
}

interface UserMetadataAction {
    type: typeof SET_USER_INFO;
    payload: UserAttributes;
}

interface UserExchangeRateAction {
    type: typeof SET_USER_EXCHANGE_RATE;
    payload: number;
}

interface UserSetBalanceAction {
    type: typeof SET_USER_WALLET_BALANCE;
    payload: string;
}

interface UserSetIsBeneficiaryAction {
    type: typeof SET_USER_IS_BENEFICIARY;
    payload: boolean;
}

interface UserSetIsBlockedAction {
    type: typeof SET_USER_IS_BLOCKED;
    payload: boolean;
}

interface UserSetIsSuspectAction {
    type: typeof SET_USER_IS_SUSPECT;
    payload: boolean;
}

interface UserSetIsCommunityManagerAction {
    type: typeof SET_USER_IS_COMMUNITY_MANAGER;
    payload: boolean;
}

interface UserSetBeneficiaryAcceptedRulesAction {
    type: typeof SET_APP_BENEFICIARY_HAS_ACCEPTED_TERMS;
    payload: boolean;
}

interface UserSetManagerAcceptedRulesAction {
    type: typeof SET_APP_MANAGER_HAS_ACCEPTED_TERMS;
    payload: boolean;
}

interface CeloKitAction {
    type: typeof SET_CELO_KIT;
    payload: ContractKit;
}

interface SetCommunityContractAction {
    type: typeof SET_COMMUNITY_CONTRACT;
    payload: any;
}

interface SetCommunityMetadataAction {
    type: typeof SET_COMMUNITY;
    payload: CommunityAttributes;
}

interface ResetUserAction {
    type: typeof RESET_USER_APP;
    payload: any;
}

interface SetTokenPushNotificationsAction {
    type: typeof SET_PUSH_NOTIFICATION_TOKEN;
    payload: string;
}

interface SetAuthTokenAction {
    type: typeof SET_AUTH_TOKEN;
    payload: string;
}
interface UserLanguageAction {
    type: typeof SET_USER_LANGUAGE;
    payload: string;
}

interface SetAppSuspectWrongDateTime {
    type: typeof SET_APP_SUSPECT_WRONG_DATETIME;
    payload: {
        suspect: boolean;
        timeDiff: number;
    };
}

interface SetAppFromWelcomeScreen {
    type: typeof SET_APP_FROM_WELCOME_SCREEN;
    payload: string;
}

interface SetAppEchangeRatesAction {
    type: typeof SET_EXCHANGE_RATES;
    payload: any;
}

interface OpenModalDonateAction {
    type: typeof modalDonateAction.OPEN;
    payload: CommunityAttributes;
}

interface GoToConfirmModalDonateAction {
    type: typeof modalDonateAction.GO_TO_CONFIRM_DONATE;
    payload: {
        inputAmount: string;
        amountInDollars: number;
        backNBeneficiaries: number;
        backForDays: number;
    };
}

interface GoToErrorModalDonateAction {
    type: typeof modalDonateAction.GO_TO_ERROR_DONATE;
}

interface GoBackToModalDonateAction {
    type: typeof modalDonateAction.GO_BACK_TO_DONATE;
}

interface InProgressModalDonateAction {
    type: typeof modalDonateAction.IN_PROGRESS;
    payload: boolean;
}

interface CloseModalDonateAction {
    type: typeof modalDonateAction.CLOSE;
}

interface InitLoadStoriesActionRequest {
    type: typeof storiesAction.INIT_REQUEST;
    payload: { start: number; end: number };
}

interface InitLoadStoriesActionSuccess {
    type: typeof storiesAction.INIT_SUCCESS;
    payload: ICommunitiesListStories[];
}

interface InitLoadStoriesActionFailure {
    type: typeof storiesAction.INIT_FAILURE;
}

interface LoadMyStoriesActionRequest {
    type: typeof storiesAction.USER_STORIES_REQUEST;
    payload: ICommunityStory[];
}

interface LoadMyStoriesActionSuccess {
    type: typeof storiesAction.USER_STORIES_SUCCESS;
    payload: ICommunityStory[];
}

interface LoadMyStoriesActionFailure {
    type: typeof storiesAction.USER_STORIES_FAILURE;
}

interface LoadMoreStoriesAction {
    type: typeof storiesAction.CONCAT;
    payload: ICommunitiesListStories[];
}

interface SetAppPushNotificationListeners {
    type: typeof appAction.SET_PUSH_NOTIFICATION_LISTENERS;
    payload: {
        notificationReceivedListener: Subscription;
        notificationResponseReceivedListener: Subscription;
    };
}

export type UserActionTypes =
    | UserWalletAction
    | UserSetBalanceAction
    | UserSetIsBeneficiaryAction
    | UserSetIsBlockedAction
    | UserSetIsSuspectAction
    | UserSetIsCommunityManagerAction
    | ResetUserAction
    | UserMetadataAction
    | UserLanguageAction
    | UserExchangeRateAction
    | SetCommunityContractAction
    | SetCommunityMetadataAction;

export type AuthActionTypes =
    | SetTokenPushNotificationsAction
    | SetAuthTokenAction;

export type AppActionTypes =
    | CeloKitAction
    | SetAppEchangeRatesAction
    | SetAppSuspectWrongDateTime
    | SetAppFromWelcomeScreen
    | SetAppPushNotificationListeners
    | UserSetBeneficiaryAcceptedRulesAction
    | UserSetManagerAcceptedRulesAction;

export type ModalActionTypes =
    | OpenModalDonateAction
    | GoToConfirmModalDonateAction
    | GoToErrorModalDonateAction
    | GoBackToModalDonateAction
    | InProgressModalDonateAction
    | CloseModalDonateAction;

export type StoriesActionTypes =
    | InitLoadStoriesActionRequest
    | InitLoadStoriesActionSuccess
    | InitLoadStoriesActionFailure
    | LoadMoreStoriesAction
    | LoadMyStoriesActionRequest
    | LoadMyStoriesActionSuccess
    | LoadMyStoriesActionFailure;

export type IStoreCombinedActionsTypes =
    | UserActionTypes
    | AuthActionTypes
    | AppActionTypes
    | ModalActionTypes
    | StoriesActionTypes;
