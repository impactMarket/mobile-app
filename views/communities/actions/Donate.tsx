import React, { Component } from 'react';
import {
    StyleSheet,
    Clipboard,
    ToastAndroid,
    Alert,
    Platform,
} from 'react-native';
import {
    ICommunityInfo, IRootState,
} from '../../../helpers/types';
import {
    Button,
    Dialog,
    Paragraph,
    Portal,
    TextInput,
} from 'react-native-paper';
import { ConnectedProps, connect } from 'react-redux';
import { celoWalletRequest } from '../../../services';
import BigNumber from 'bignumber.js';


interface IExploreScreenProps {
    community: ICommunityInfo
}
const mapStateToProps = (state: IRootState) => {
    const { user, network } = state
    return { user, network }
};

const connector = connect(mapStateToProps)

type PropsFromRedux = ConnectedProps<typeof connector>

type Props = PropsFromRedux & IExploreScreenProps
interface IDonateState {
    openModalDonate: boolean;
    openModalDonateWithCelo: boolean;
    donating: boolean;
    amountDonate: string;
}
class Donate extends Component<Props, IDonateState> {
    constructor(props: any) {
        super(props);
        this.state = {
            openModalDonate: false,
            openModalDonateWithCelo: false,
            donating: false,
            amountDonate: '',
        }
    }

    handleOpenDonateWithCelo = () => {
        this.setState({ openModalDonate: false, openModalDonateWithCelo: true });
    }

    handleDonateWithCeloWallet = async () => {
        this.setState({ donating: true });
        const stableToken = await this.props.network.kit.contracts.getStableToken()
        const cUSDDecimals = await stableToken.decimals();
        const txObject = stableToken.transfer(
            this.props.community.contractAddress,
            new BigNumber(
                this.state.amountDonate
            ).multipliedBy(
                new BigNumber(10).pow(cUSDDecimals)
            ).toString()
        ).txo
        const requestId = "donate-to-community";
        celoWalletRequest(
            this.props.user.celoInfo.address,
            stableToken.address,
            txObject,
            requestId,
            this.props.network,
        ).then(() => {
            Alert.alert(
                'Success',
                'You\'ve donated!',
                [
                    { text: 'OK' },
                ],
                { cancelable: false }
            );
        }).catch(() => {
            Alert.alert(
                'Failure',
                'An error happened while donating!',
                [
                    { text: 'OK' },
                ],
                { cancelable: false }
            );
        }).finally(() => {
            this.setState({
                openModalDonate: false,
                openModalDonateWithCelo: false,
                donating: false,
                amountDonate: '',
            });
        });
    }

    handleCopyAddressToClipboard = () => {
        Clipboard.setString(this.props.community.contractAddress)
        if (Platform.OS === 'android') {
            ToastAndroid.show('Address copied to clipboard!', ToastAndroid.SHORT);
        }
        // TODO: add ios notification
    }

    render() {
        const {
            openModalDonate,
            openModalDonateWithCelo,
            donating,
            amountDonate,
        } = this.state;
        const {
            community,
        } = this.props;

        return (
            <>
                <Button
                    mode="contained"
                    onPress={() => this.setState({ openModalDonate: true })}
                >
                    Donate
                </Button>
                <Portal>
                    <Dialog
                        visible={openModalDonate}
                        onDismiss={() => this.setState({ openModalDonate: false })}
                    >
                        <Dialog.Title>Donate</Dialog.Title>
                        <Dialog.Content>
                            <Paragraph>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla lacinia eros ut tortor rhoncus fringilla.
                            </Paragraph>
                            <Paragraph style={{ marginTop: 20 }}>
                                Donating to {community.name}
                            </Paragraph>
                            <Button
                                mode="contained"
                                onPress={this.handleCopyAddressToClipboard}
                            >
                                {community.contractAddress.substr(0, 10)}...{community.contractAddress.substr(community.contractAddress.length - 7, community.contractAddress.length)}
                            </Button>
                            <Button
                                mode="contained"
                                disabled={this.props.user.celoInfo.address.length === 0}
                                style={{ marginTop: 10 }}
                                onPress={this.handleOpenDonateWithCelo}
                            >
                                Donate with Celo wallet
                            </Button>
                        </Dialog.Content>
                        <Dialog.Actions>
                            <Button mode="contained" onPress={() => this.setState({ openModalDonate: false })}>Done</Button>
                        </Dialog.Actions>
                    </Dialog>
                    <Dialog
                        visible={openModalDonateWithCelo}
                        onDismiss={() => this.setState({ openModalDonateWithCelo: false })}
                    >
                        <Dialog.Title>Donate</Dialog.Title>
                        <Dialog.Content>
                            <Paragraph>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla lacinia eros ut tortor rhoncus fringilla.
                            </Paragraph>
                            <TextInput
                                label='Amount ($)'
                                mode="outlined"
                                keyboardType="numeric"
                                value={amountDonate}
                                onChangeText={text => this.setState({ amountDonate: text })}
                            />
                        </Dialog.Content>
                        <Dialog.Actions>
                            <Button mode="contained" loading={donating} style={{ marginRight: 10 }} onPress={this.handleDonateWithCeloWallet}>Donate</Button>
                            <Button mode="contained" onPress={() => this.setState({ openModalDonateWithCelo: false })}>Close</Button>
                        </Dialog.Actions>
                    </Dialog>
                </Portal>
            </>
        );
    }
}

const styles = StyleSheet.create({
});

export default connector(Donate)