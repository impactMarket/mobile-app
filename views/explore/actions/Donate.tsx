import React, { useState } from 'react';
import {
    StyleSheet,
    Clipboard,
    ToastAndroid,
} from 'react-native';
import {
    ICommunityViewInfo,
} from '../../../helpers/types';
import { Button, Dialog, Paragraph, Portal, TextInput } from 'react-native-paper';


export default function Donate(props: { community: ICommunityViewInfo }) {
    const [openModalDonate, setOpenModalDonate] = useState(false);
    const [openModalDonateWithCelo, setOpenModalDonateWithCelo] = useState(false);
    const [amountDonate, setAmountDonate] = React.useState('');
    const { community } = props;

    const handleOpenDonateWithCelo = () => {
        setOpenModalDonateWithCelo(true);
        setOpenModalDonate(false);
    }

    const handleDonateWithCeloWallet = () => {
        // TODO: open modal to introduce amount and then send request
    }

    const handleCopyAddressToClipboard = () => {
        Clipboard.setString(community.contractAddress)
        ToastAndroid.show('Address copied to clipboard!', ToastAndroid.SHORT);
    }

    return (
        <>
            <Button
                mode="contained"
                onPress={() => setOpenModalDonate(true)}
            >
                Donate
            </Button>
            <Portal>
                <Dialog
                    visible={openModalDonate}
                    onDismiss={() => setOpenModalDonate(false)}
                >
                    <Dialog.Title>Donate</Dialog.Title>
                    <Dialog.Content>
                        <Paragraph>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla lacinia eros ut tortor rhoncus fringilla.
                        </Paragraph>
                        <Paragraph style={{ marginTop: 20 }}>
                            Donating to {community.name}
                        </Paragraph>
                        <Button
                            mode="contained"
                            onPress={handleCopyAddressToClipboard}
                        >
                            {community.contractAddress.substr(0, 10)}...{community.contractAddress.substr(community.contractAddress.length - 7, community.contractAddress.length)}
                        </Button>
                        <Button
                            mode="contained"
                            style={{ marginTop: 10 }}
                            onPress={handleOpenDonateWithCelo}
                        >
                            Donate with Celo wallet
                        </Button>
                    </Dialog.Content>
                    <Dialog.Actions>
                        <Button mode="contained" onPress={() => setOpenModalDonate(false)}>Done</Button>
                    </Dialog.Actions>
                </Dialog>
                <Dialog
                    visible={openModalDonateWithCelo}
                    onDismiss={() => setOpenModalDonateWithCelo(false)}
                >
                    <Dialog.Title>Donate</Dialog.Title>
                    <Dialog.Content>
                        <Paragraph>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla lacinia eros ut tortor rhoncus fringilla.
                        </Paragraph>
                        <TextInput
                            label='Amount'
                            mode="outlined"
                            keyboardType="numeric"
                            value={amountDonate}
                            onChangeText={text => setAmountDonate(text)}
                        />
                    </Dialog.Content>
                    <Dialog.Actions>
                        <Button mode="contained" style={{ marginRight: 10 }} onPress={handleDonateWithCeloWallet}>Donate</Button>
                        <Button mode="contained" onPress={() => setOpenModalDonateWithCelo(false)}>Close</Button>
                    </Dialog.Actions>
                </Dialog>
            </Portal>
        </>
    );
}

const styles = StyleSheet.create({
});