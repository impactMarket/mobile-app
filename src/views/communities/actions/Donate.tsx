import i18n from 'assets/i18n';
import BigNumber from 'bignumber.js';
import {
    iptcColors,
    updateCommunityInfo,
    formatInputAmountToTransfer,
    getCurrencySymbol,
} from 'helpers/index';
import { ICommunityInfo, IRootState } from 'helpers/types';
import React, { Component } from 'react';
import { StyleSheet, Clipboard, Alert, View, TextInput } from 'react-native';
import {
    // Button,
    Paragraph,
    Portal,
    Snackbar,
    Text,
    Modal,
    Headline,
    IconButton,
    // Card,
} from 'react-native-paper';
import { ConnectedProps, connect } from 'react-redux';
import { analytics } from 'services/analytics';
import { celoWalletRequest } from 'services/celoWallet';
import { writeLog } from 'services/logger/write';
import config from '../../../../config';
import * as Device from 'expo-device';
import Button from 'components/Button';
import Card from 'components/Card';

interface IExploreScreenProps {
    community: ICommunityInfo;
}
const mapStateToProps = (state: IRootState) => {
    const { user, app } = state;
    return { user, app };
};

const connector = connect(mapStateToProps);
type PropsFromRedux = ConnectedProps<typeof connector>;
type Props = PropsFromRedux & IExploreScreenProps;

interface IDonateState {
    openModalDonate: boolean;
    donating: boolean;
    amountDonate: string;
    showCopiedToClipboard: boolean;
    modalConfirmSend: boolean;
    rates: any;
}
BigNumber.config({ EXPONENTIAL_AT: [-7, 30] });
class Donate extends Component<Props, IDonateState> {
    constructor(props: any) {
        super(props);
        this.state = {
            openModalDonate: false,
            donating: false,
            amountDonate: '',
            showCopiedToClipboard: false,
            modalConfirmSend: false,
            rates: this.props.app.exchangeRates,
        };
    }

    handleConfirmDonateWithCeloWallet = () => {
        const { amountDonate } = this.state;
        const amountInDollars =
            parseFloat(formatInputAmountToTransfer(amountDonate)) /
            this.props.user.user.exchangeRate;
        if (
            amountInDollars >
            new BigNumber(this.props.user.celoInfo.balance)
                .dividedBy(10 ** config.cUSDDecimals)
                .toNumber()
        ) {
            Alert.alert(
                i18n.t('donate'),
                i18n.t('donationBiggerThanBalance'),
                [{ text: i18n.t('close') }],
                { cancelable: true }
            );
            return;
        }
        const { community, user } = this.props;
        Alert.alert(
            i18n.t('donate'),
            i18n.t('donateConfirmMessage', {
                symbol: getCurrencySymbol(user.user.currency),
                amount: amountDonate.replace(/\B(?=(\d{3})+(?!\d))/g, ','),
                amountInDollars: amountInDollars.toFixed(2),
                to: community.name,
            }),
            [
                {
                    text: i18n.t('donate'),
                    onPress: () => this.donateWithCeloWallet(),
                },
                {
                    text: i18n.t('cancel'),
                },
            ],
            { cancelable: true }
        );
    };

    donateWithCeloWallet = async () => {
        this.setState({ donating: true });
        const stableToken = await this.props.app.kit.contracts.getStableToken();
        const cUSDDecimals = await stableToken.decimals();
        const amountInDollars =
            parseFloat(formatInputAmountToTransfer(this.state.amountDonate)) /
            this.props.user.user.exchangeRate;
        const txObject = stableToken.transfer(
            this.props.community.contractAddress,
            new BigNumber(amountInDollars)
                .multipliedBy(new BigNumber(10).pow(cUSDDecimals))
                .toString()
        ).txo;
        celoWalletRequest(
            this.props.user.celoInfo.address,
            stableToken.address,
            txObject,
            'donatetocommunity',
            this.props.app.kit
        )
            .then((tx) => {
                console.log(tx);
                // TODO: wait for tx confirmation and request UI update
                // update donated values
                setTimeout(
                    () =>
                        updateCommunityInfo(
                            this.props.user.celoInfo.address,
                            this.props
                        ),
                    10000
                );

                Alert.alert(
                    i18n.t('success'),
                    i18n.t('youHaveDonated'),
                    [{ text: 'OK' }],
                    { cancelable: false }
                );
                analytics('donate', { device: Device.brand, success: true });
            })
            .catch((e) => {
                writeLog({ action: 'donate', details: e.message });
                analytics('donate', { device: Device.brand, success: false });
                Alert.alert(
                    i18n.t('failure'),
                    i18n.t('errorDonating'),
                    [{ text: 'OK' }],
                    { cancelable: false }
                );
            })
            .finally(() => {
                this.setState({
                    openModalDonate: false,
                    donating: false,
                    amountDonate: '',
                });
            });
    };

    handleCopyAddressToClipboard = () => {
        Clipboard.setString(this.props.community.contractAddress);
        this.setState({ showCopiedToClipboard: true });
        setTimeout(() => this.setState({ showCopiedToClipboard: false }), 5000);
        this.setState({ openModalDonate: false });
    };

    render() {
        const {
            openModalDonate,
            donating,
            amountDonate,
            showCopiedToClipboard,
            rates,
        } = this.state;
        const { community, user } = this.props;

        const amountInDollars =
            parseFloat(formatInputAmountToTransfer(amountDonate)) /
            this.props.user.user.exchangeRate;
        const amountInCommunityCurrency =
            amountInDollars * rates[community.currency].rate;

        let donatingModalString = '';
        const backForDays =
            amountInDollars /
            new BigNumber(community.vars._claimAmount)
                .dividedBy(10 ** config.cUSDDecimals)
                .toNumber() /
            community.beneficiaries.added.length;
        if (backForDays < 1) {
            donatingModalString = i18n.t('yourDonationWillBack', {
                backNBeneficiaries: Math.max(
                    1,
                    Math.floor(
                        amountInDollars /
                            new BigNumber(community.vars._claimAmount)
                                .dividedBy(10 ** config.cUSDDecimals)
                                .toNumber()
                    )
                ),
            });
        } else {
            donatingModalString = i18n.t('yourDonationWillBackFor', {
                backNBeneficiaries: community.beneficiaries.added.length,
                backForDays: Math.floor(backForDays),
            });
        }

        const donateWithValoraButton =
            user.celoInfo.address.length > 0 ? (
                <Button
                    modeType="default"
                    bold={true}
                    loading={donating}
                    disabled={donating}
                    onPress={this.handleConfirmDonateWithCeloWallet}
                >
                    {i18n.t('donateWithValora')}
                </Button>
            ) : (
                <Button
                    icon="alert"
                    modeType="default"
                    bold={true}
                    style={{
                        backgroundColor: '#f0ad4e',
                    }}
                    onPress={() => {
                        Alert.alert(
                            i18n.t('failure'),
                            i18n.t('youAreNotConnected'),
                            [{ text: i18n.t('close') }],
                            { cancelable: false }
                        );
                    }}
                >
                    {i18n.t('donateWithValora')}
                </Button>
            );

        return (
            <>
                <Button
                    modeType="green"
                    bold={true}
                    style={styles.donate}
                    labelStyle={{
                        fontSize: 20,
                        lineHeight: 23,
                        color: 'white',
                        fontWeight: 'bold',
                        // backgroundColor: 'red',
                    }}
                    onPress={() => this.setState({ openModalDonate: true })}
                >
                    {i18n.t('donate')}
                </Button>
                <Snackbar
                    visible={showCopiedToClipboard}
                    onDismiss={() => {}}
                    action={{
                        label: '',
                        onPress: () => {
                            // Do something
                        },
                    }}
                >
                    {i18n.t('addressCopiedClipboard')}
                </Snackbar>
                <Portal>
                    <Modal
                        visible={openModalDonate}
                        onDismiss={() =>
                            this.setState({ openModalDonate: false })
                        }
                    >
                        <Card style={{ marginHorizontal: 20 }}>
                            <Card.Content>
                                <View style={{ height: 40 }}>
                                    <View
                                        style={{
                                            flex: 1,
                                            flexDirection: 'row',
                                            justifyContent: 'space-between',
                                        }}
                                    >
                                        <Headline style={{ marginVertical: 3 }}>
                                            {i18n.t('donateSymbol', {
                                                symbol: user.user.currency,
                                            })}
                                        </Headline>
                                        <IconButton
                                            icon="close"
                                            size={20}
                                            style={{ top: -5 }}
                                            onPress={() =>
                                                this.setState({
                                                    openModalDonate: false,
                                                })
                                            }
                                        />
                                    </View>
                                </View>
                                <View
                                    style={{
                                        backgroundColor: '#F6F6F7',
                                        // opacity: 0.27,
                                        alignItems: 'center',
                                        borderRadius: 5,
                                    }}
                                >
                                    <TextInput
                                        keyboardType="numeric"
                                        placeholder={i18n.t('amountSymbol', {
                                            symbol: getCurrencySymbol(
                                                user.user.currency
                                            ),
                                        })}
                                        style={{
                                            fontFamily: 'Gelion-Regular',
                                            fontStyle: 'normal',
                                            fontWeight: 'normal',
                                            fontSize: 40,
                                            lineHeight: 50,
                                            textAlign: 'center',
                                        }}
                                        value={amountDonate}
                                        onChangeText={(text) =>
                                            this.setState({
                                                amountDonate: text,
                                            })
                                        }
                                    />
                                    {amountDonate.length > 0 && (
                                        <Paragraph
                                            style={{ marginVertical: 10 }}
                                        >
                                            ~
                                            {`${getCurrencySymbol(
                                                community.currency
                                            )}${
                                                (Math.floor(
                                                    amountInCommunityCurrency *
                                                        100
                                                ) / 100).toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
                                            } ${community.currency}`}
                                        </Paragraph>
                                    )}
                                </View>
                                {amountDonate.length > 0 && (
                                    <Paragraph
                                        style={{
                                            marginBottom: 20,
                                            textAlign: 'center',
                                        }}
                                    >
                                        {donatingModalString}
                                    </Paragraph>
                                )}
                                <Button
                                    modeType="gray"
                                    bold={true}
                                    style={{
                                        // backgroundColor: '#F2F3F5',
                                        marginVertical: 10,
                                    }}
                                    onPress={this.handleCopyAddressToClipboard}
                                >
                                    {i18n.t('copyContractAddress')}
                                </Button>
                                {donateWithValoraButton}
                            </Card.Content>
                        </Card>
                    </Modal>
                </Portal>
            </>
        );
    }
}

const styles = StyleSheet.create({
    donate: {
        borderRadius: 0,
        height: 69,
    },
});

export default connector(Donate);
