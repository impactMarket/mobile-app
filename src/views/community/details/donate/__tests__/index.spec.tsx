import combinedReducer from 'helpers/redux/reducers';
import React from 'react';
import { Provider } from 'react-redux';
import { create } from 'react-test-renderer';
import { createStore } from 'redux';

import Donate from '../index';

const store = createStore(combinedReducer);

const mockedCommunityAttributes = {
    id: 2,
    publicId: '5',
    requestByAddress: '',
    contractAddress: '',
    name: 'Communuity Test',
    description: 'Communuity Test',
    descriptionEn: null,
    language: 'en',
    currency: 'USD',
    city: 'Lisbon',
    country: 'Portugal',
    gps: {
        latitude: 123,
        longitude: 123,
    },
    email: 'teste@teste.com',
    visibility: 'public',
    coverMediaId: 4,
    status: 'valid',
    started: new Date(),

    // timestamps
    createdAt: new Date(),
    updatedAt: new Date(),
};

const tree = create(
    <Provider store={store}>
        <Donate community={mockedCommunityAttributes} />
    </Provider>
).toJSON();

describe('CommunityManager screen test', () => {
    test('should render Donate correctly', () => {
        expect(tree).toMatchSnapshot();
    });
});
