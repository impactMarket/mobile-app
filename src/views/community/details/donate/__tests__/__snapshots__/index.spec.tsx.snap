// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommunityManager screen test should render Donate correctly 1`] = `
<View
  bold={true}
  modeType="green"
  style={
    Object {
      "backgroundColor": "#2DCE89",
      "borderColor": "transparent",
      "borderRadius": 0,
      "borderStyle": "solid",
      "borderWidth": 0,
      "elevation": 0,
      "height": 69,
      "justifyContent": "center",
      "minWidth": 64,
    }
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 4,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          },
          undefined,
        ]
      }
    >
      <Text
        numberOfLines={1}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "letterSpacing": 1,
                "marginHorizontal": 16,
                "marginVertical": 9,
                "textAlign": "center",
              },
              undefined,
              false,
              Object {
                "color": "#ffffff",
                "fontFamily": "System",
                "fontWeight": "500",
              },
              Object {
                "fontFamily": "System",
                "fontWeight": "500",
              },
              Object {
                "color": "white",
                "fontFamily": "Inter-Bold",
                "fontSize": 20,
                "lineHeight": 23,
              },
            ],
          ]
        }
      >
        Contribute
      </Text>
    </View>
  </View>
</View>
`;
